### PR TITLE
[yaml] Adding Spanner IO Providers for Beam YAML

### PR DESCRIPTION
**Pull request by reeba212**
1. Added _**SpannerReadSchemaTransformProvider.java**_, a provider for reading from spanner.
2. Modified **_SpannerWriteSchemaTransformProvider.java_** to incorporate error handling and fixed other issues.
3. Modified **_MutationUtils.java_** to add a function "createRowFromMutation" which creates a Beam Row from a Mutation. This is used in the spanner write provider.
4. Modified **_ErrorHandling.java_** to make it serializable.
5. Modified **_standard_io.yaml_** to register the Spanner IO providers.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/yaml/standard_io.yaml
Comment (Polber) [inline sdks/python/apache_beam/yaml/standard_io.yaml:277]: I would strip the `_id` from all of these parameters on the YAML side

i.e.
```suggestion
        project: 'project_id'
        instance: 'instance_id'
        database: 'database_id'
        table: 'table_id'
        query: 'query'
        columns: 'columns'
      'WriteToSpanner':
        instance: 'instance_id'
        database: 'database_id'
        table: 'table_id'
```

**Review by Polber — COMMENTED** [review]
Thanks Reeba! I left some comments, but the overall structure looks really good

**Comment by Polber** [comment]
R: @damccorm

File: sdks/python/apache_beam/io/gcp/spanner_wrapper.py
Comment (Polber) [inline sdks/python/apache_beam/io/gcp/spanner_wrapper.py:34]: I have no idea why they wrote the BQ wrapper this way, but these don't need to be class variables as they are state variables (except `TEMP_DATABASE_PREFIX`, I suppose) I would remove at least the state variables and move to constructor as instance variables.
```suggestion
  TEMP_DATABASE_PREFIX = 'temp-'
```

File: sdks/python/apache_beam/io/gcp/spanner_wrapper.py
Comment (Polber) [inline sdks/python/apache_beam/io/gcp/spanner_wrapper.py:39]: Refactoring their usages in the rest of the class accordingly 
```suggestion
    self._spanner_client = spanner.Client(project = project_id)
    self._spanner_instance = self._spanner_client.instance("beam-test")
    self._test_database = None
```

File: sdks/python/apache_beam/io/gcp/spanner_wrapper.py
Comment (Polber) [inline sdks/python/apache_beam/io/gcp/spanner_wrapper.py:69]: This isn't truly acting as a class method as it is using state of object (class methods are useful for methods that are used independently from the state of the class, for example a method that returns every created table across all instances of the object)
```suggestion
```

and for future reference, a classmethod follows the standard 
```
@classmethod
def some_method(cls):
  return cls.some_var
```
using `cls` instead of `self` to access class variables

**Review by Polber — CHANGES_REQUESTED** [review]
Added some comments related to newly added testing framework

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java
Comment (Polber) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java:162]: This will never fail since `Strings.isNullOrEmpty(this.getQuery())` is a boolean (never null)
```suggestion
        checkNotNull(Strings.isNullOrEmpty(this.getQuery()), invalidConfigMessage + "Query must be specified for query read.");
```
```suggestion
        checkArgument(
          Strings.isNullOrEmpty(this.getQuery()),
          invalidConfigMessage + "Query must be specified for query read."
        );
```

File: sdks/python/apache_beam/yaml/integration_tests.py
Comment (Polber) [reply sdks/python/apache_beam/yaml/integration_tests.py:59]:
```suggestion
    table = 'tmp_table'
```

**Review by Polber — CHANGES_REQUESTED** [review]
Couple more small changes. Also, make sure to address the unresolved conversations. Thanks!

**Comment by Polber** [comment]
@reeba212 #32008 was merged, so you should be able to rebase on main and make sure checks pass!

File: sdks/python/apache_beam/yaml/integration_tests.py
Comment (Polber) [reply sdks/python/apache_beam/yaml/integration_tests.py]: This is a mistake on my part - leaving as 'beam-test' should be fine. Thanks!

File: sdks/python/apache_beam/io/gcp/spanner_wrapper.py
Comment (Polber) [inline sdks/python/apache_beam/io/gcp/spanner_wrapper.py:20]:
```suggestion
```

File: sdks/python/apache_beam/yaml/integration_tests.py
Comment (Polber) [inline sdks/python/apache_beam/yaml/integration_tests.py:55]:
```suggestion
```

**Review by Polber — COMMENTED** [review]
Lint errors:
```
************* Module apache_beam.io.gcp.spanner_wrapper	
apache_beam/io/gcp/spanner_wrapper.py:20:0: W0611: Unused apache_beam imported as beam (unused-import)	
************* Module apache_beam.yaml.integration_tests	
apache_beam/yaml/integration_tests.py:50:0: W0404: Reimport 'contextlib' (imported line 20) (reimported)	
apache_beam/yaml/integration_tests.py:50:0: C0413: Import "import contextlib" should be placed at the top of the module (wrong-import-position)	
apache_beam/yaml/integration_tests.py:51:0: W0404: Reimport 'uuid' (imported line 27) (reimported)	
apache_beam/yaml/integration_tests.py:51:0: C0413: Import "import uuid" should be placed at the top of the module (wrong-import-position)	
apache_beam/yaml/integration_tests.py:52:0: W0404: Reimport 'logging' (imported line 24) (reimported)	
apache_beam/yaml/integration_tests.py:52:0: C0413: Import "import logging" should be placed at the top of the module (wrong-import-position)	
apache_beam/yaml/integration_tests.py:53:0: C0413: Import "from google.cloud import spanner" should be placed at the top of the module (wrong-import-position)	
apache_beam/yaml/integration_tests.py:50:0: C0412: Imports from package contextlib are not grouped (ungrouped-imports)	
apache_beam/yaml/integration_tests.py:51:0: C0412: Imports from package uuid are not grouped (ungrouped-imports)	
apache_beam/yaml/integration_tests.py:52:0: C0412: Imports from package logging are not grouped (ungrouped-imports)	
apache_beam/yaml/integration_tests.py:53:0: W0611: Unused spanner imported from google.cloud (unused-import)
```

Make modification I added in comments then run:
`yapf --in-place --parallel --recursive apache_beam`

I also recommend following steps in https://cwiki.apache.org/confluence/display/BEAM/Python+Tips#PythonTips-LintandFormattingChecks

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java
Comment (Polber) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java:199]:
```suggestion
    public void validate() {
      String invalidConfigMessage = "Invalid Cloud Spanner Read configuration: ";
      checkArgument(
          !Strings.isNullOrEmpty(this.getInstanceId()),
          invalidConfigMessage + "Instance ID must be specified.");
      checkArgument(
          !Strings.isNullOrEmpty(this.getDatabaseId()),
          invalidConfigMessage + "Database ID must be specified.");

      if (Strings.isNullOrEmpty(this.getQuery())) {
       checkArgument(
            !Strings.isNullOrEmpty(this.getTableId()),
            invalidConfigMessage + "Table name must be specified for table read.");
        checkArgument(
            this.getColumns() != null && !this.getColumns().isEmpty(),
            invalidConfigMessage + "Columns must be specified for table read.");

      } else {
        checkArgument(
            !Strings.isNullOrEmpty(this.getQuery()),
            invalidConfigMessage + "Query must be specified for query read.");
        checkArgument(
            Strings.isNullOrEmpty(this.getTableId()),
            invalidConfigMessage + "Table name should not be specified when using a query.");
        checkArgument(
            this.getColumns() == null || this.getColumns().isEmpty(),
            invalidConfigMessage + "Columns should not be specified when using a query.");
      }
    }
```

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java
Comment (Polber) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java:207]:
```suggestion
    @SchemaFieldDescription("Specifies the GCP project ID.")
    @Nullable
    public abstract String getProjectId();
```

**Review by Polber — CHANGES_REQUESTED** [review]
Try seeing if making these changes fixes test errors

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteSchemaTransformProvider.java
Comment (Polber) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerWriteSchemaTransformProvider.java:294]: ProjectId is not required - probably why the IT is failing to build transform since it does not pass project
```suggestion
```

**Review by Polber — CHANGES_REQUESTED** [review]

File: sdks/python/apache_beam/io/gcp/spanner_wrapper_test.py
Comment (Polber) [inline sdks/python/apache_beam/io/gcp/spanner_wrapper_test.py:36]: Try just creating an __init__ method instead. (`@mock.patch` is intended to annotate a class that implements the one being mock - which is not what is happening here)
```suggestion
  def __init__(self):
    self.wrapper = SpannerWrapper(project_id='test-project')
    self.mock_instance = MagicMock()
    self.mock_database = MagicMock()
    self.mock_client.instance.return_value = self.mock_instance
    self.mock_instance.database.return_value = self.mock_database
    self.wrapper._spanner_client = mock.Mock()
    self.wrapper._spanner_instance = self.mock_instance
    self.wrapper._test_database = self.mock_database
```

File: sdks/python/setup.py
Comment (damccorm) [inline sdks/python/setup.py:447]: @Polber @reeba212 catching up on this, which suites were failing such that we needed to add this? I'd like to follow up with the spanner folks to make sure they're aware of the breaking change

**Review by damccorm — COMMENTED** [review]
Reran the failing jobs to see if we can get them green, but they do seem unrelated to this change

**Review by Polber — APPROVED** [review]
Thanks!

File: sdks/python/setup.py
Comment (reeba212) [reply sdks/python/setup.py:447]: There were two checks that were failing and both had the same underlying error: 
`ModuleNotFoundError: No module named 'grpc_interceptor'`

**Review by damccorm — APPROVED** [review]
Thanks!

**Comment by chamikaramj** [comment]
Seems like the google-cloud-spanner version specified here conflicts with the version installed in the containers.

https://github.com/apache/beam/blob/master/sdks/python/setup.py#L444
https://github.com/apache/beam/blob/master/sdks/python/container/py38/base_image_requirements.txt#L72

At least one test suite is failing due to this.
https://ge.apache.org/s/2eh6tfikgiwb2/console-log/task/:sdks:python:expansion-service-container:docker?anchor=258&page=1

cc: @tvalentyn

**Comment by liferoad** [comment]
cc @Polber

**Comment by Polber** [comment]
@damccorm was working on rebuilding those - not sure the status

**Comment by Polber** [comment]
Looks like it is under review still: #32411

**Comment by damccorm** [comment]
Yep, was just waiting on approval - just merged and it should be back to green

**Pull request merged**
